<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>

    <script>
        function execCode(counter) {
            // 定义一个promise对象
            const obj = new Promise((resolve, reject) => {
                // 这里的代码在创建时会立即执行 
                // 一般将异步的代码放在回调函数里
                setTimeout(() => {
                    console.log("hello world")
                    // 特殊情况 
                    if(counter <= 0) {
                        // 失败回调
                        // obj.failure(`${counter}有问题`)
                        reject(`${counter}有问题`)

                    } else {
                        let count = 0
                        for(let i = 0; i < 100; i++) {
                            count += i
                        }
                        // 成功回调
                        // obj.success(count)
                        resolve(count)

                    }
                }, 3000)
            })
            

            // 返回一个凭证
            return obj
        }

        const promise = execCode(10)
        // 监听返回值
        promise.then((value) => {
            console.log("执行完了", value)
        })
        promise.catch((err) => {
            console.log("执行失败", err)
        })
        const promise1 = execCode(-10)
        // 监听返回值
        promise1.then((value) => {
            console.log("执行完了", value)
        })
        promise1.catch((err) => {
            console.log("执行失败", err)
        })

        // 简洁写法 ***
        execCode(100).then((value) => {
            console.log("执行完了", value)
        }).catch((err) => {
            console.log("执行失败", err)
        })

        console.log('----------------')
        // Executor是在创建promise时需要传入的一个回调函数 这个回调函数会被立即执行 并且传入两个参数
        const promise2 = new Promise((resolve, reject) => {
            // 创建完promise
            // 1.待定状态（pending），初始状态 既没有被兑现，也没有被拒绝
            // 开始执行execCode时，处于这种状态
            console.log("待定状态")
            // 这两个函数都会回调
            // 2.已兑现（fulfilled），执行了resolve时，处于该状态
            // resolve(1)
            resolve([
                {name: "zhang", age: 20},
                {name: "li", age: 18},
            ])
            // 3.已拒绝（rejected），执行了reject时处于该状态
            reject()
            // promise的状态一旦确定 就不会被更改
            // 也不能再执行某一个回调函数来改变状态
            
            

        }) 
        // 
        promise2.then((res) => {
            console.log("执行成功", res)
        })
        promise2.catch((err) => {
            console.log("执行失败", err)
        })

        console.log('---------------')
        const promise3 = new Promise((resolve, reject) => {
            // resolve("success")
            reject("failure")
        })
        // then参数的传递方式 ***
        promise3.then((res) => {
            console.log("执行成功", res)
        }, (err) => {
            console.log("执行失败", err)
        })
       


    </script>
    
</body>
</html>